File: agent/src/research_agent/contracts.py
"""Loaders for the cross-language contracts in `contracts/`.

The YAML files are the single source of truth shared with the Go dispatcher; the enums below are
the statically typed Python view of them. `tests/unit/test_contracts.py` asserts that the two stay
in sync in both directions, so adding a code on one side only is a test failure, not a surprise in
production (architecture v0.6 §2, §13).
"""

from __future__ import annotations

from enum import StrEnum
from functools import cache
from typing import Any, Self

import yaml
from pydantic import BaseModel, ConfigDict, Field, model_validator

from research_agent.paths import contracts_dir


class RunStatus(StrEnum):
    """Run lifecycle states (v0.6 §2)."""

    QUEUED = "queued"
    DISPATCHED = "dispatched"
    RUNNING = "running"
    SUCCEEDED = "succeeded"
    SUCCEEDED_WITH_WARNINGS = "succeeded_with_warnings"
    FAILED = "failed"
    CANCELLED = "cancelled"


class StateSpec(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    owner: str
    transitions: tuple[RunStatus, ...]


class RunStateMachine(BaseModel):
    """The allowed run state transitions, as declared by `contracts/run_states.yaml`."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    version: int
    initial: RunStatus
    terminal: frozenset[RunStatus]
    max_attempts: int = Field(ge=1)
    states: dict[RunStatus, StateSpec]

    @model_validator(mode="after")
    def _check_consistency(self) -> Self:
        missing = set(RunStatus) - set(self.states)
        if missing:
            raise ValueError(f"states missing from the contract: {sorted(missing)}")
        for status in self.terminal:
            if self.states[status].transitions:
                raise ValueError(f"terminal state {status} declares transitions")
        if self.initial in self.terminal:
            raise ValueError(f"initial state {self.initial} cannot be terminal")
        return self

    def can_transition(self, source: RunStatus, target: RunStatus) -> bool:
        return target in self.states[source].transitions

    def is_terminal(self, status: RunStatus) -> bool:
        return status in self.terminal


class ErrorCodeSpec(BaseModel):
    """Metadata for one error code.

    `expected` separates handled degradation from bugs: expected errors are logged at warn level
    with the error -> decision -> outcome chain, unexpected ones fail the run with a stack trace
    and a searchable event id (v0.6 §13.1).
    """

    model_config = ConfigDict(frozen=True, extra="forbid")

    category: str
    expected: bool
    retryable: bool
    description: str


def _load_yaml(name: str) -> dict[str, Any]:
    with (contracts_dir() / name).open(encoding="utf-8") as handle:
        loaded: dict[str, Any] = yaml.safe_load(handle)
    return loaded


@cache
def run_state_machine() -> RunStateMachine:
    return RunStateMachine.model_validate(_load_yaml("run_states.yaml"))


@cache
def error_code_specs() -> dict[str, ErrorCodeSpec]:
    raw = _load_yaml("error_codes.yaml")
    return {name: ErrorCodeSpec.model_validate(spec) for name, spec in sorted(raw["codes"].items())}

File: dispatcher/internal/contracts/contracts.go
// Package contracts loads the files in contracts/ that Go shares with Python: the run state
// machine and the error code taxonomy. The dispatcher refuses to start if either is missing or
// inconsistent, and every status write it makes is checked against the state machine first.
package contracts

import (
	"fmt"
	"os"
	"path/filepath"
	"slices"

	"gopkg.in/yaml.v3"
)

// Status mirrors RunStatus in agent/src/research_agent/contracts.py.
type Status string

const (
	Queued                Status = "queued"
	Dispatched            Status = "dispatched"
	Running               Status = "running"
	Succeeded             Status = "succeeded"
	SucceededWithWarnings Status = "succeeded_with_warnings"
	Failed                Status = "failed"
	Cancelled             Status = "cancelled"
)

// AllStatuses is the full vocabulary; a contract test asserts it equals the YAML.
var AllStatuses = []Status{
	Queued, Dispatched, Running, Succeeded, SucceededWithWarnings, Failed, Cancelled,
}

// Error codes the dispatcher itself produces. Each must exist in error_codes.yaml.
const (
	CodeAgentUnreachable  = "AGENT_UNREACHABLE"
	CodeHeartbeatLost     = "AGENT_HEARTBEAT_LOST"
	CodeDeadlineExceeded  = "DEADLINE_EXCEEDED"
	CodeDispatchDeferred  = "DISPATCH_DEFERRED"
	CodeUnexpectedFailure = "UNEXPECTED_EXCEPTION"
)

// CodesUsed is checked against the shared taxonomy at startup and in tests.
var CodesUsed = []string{
	CodeAgentUnreachable, CodeHeartbeatLost, CodeDeadlineExceeded, CodeDispatchDeferred,
	CodeUnexpectedFailure,
}

type stateSpec struct {
	Owner       string   `yaml:"owner"`
	Transitions []Status `yaml:"transitions"`
}

// StateMachine is contracts/run_states.yaml.
type StateMachine struct {
	Version     int                  `yaml:"version"`
	Initial     Status               `yaml:"initial"`
	Terminal    []Status             `yaml:"terminal"`
	MaxAttempts int                  `yaml:"max_attempts"`
	States      map[Status]stateSpec `yaml:"states"`
}

// CanTransition reports whether source -> target is allowed.
func (m StateMachine) CanTransition(source, target Status) bool {
	spec, ok := m.States[source]
	return ok && slices.Contains(spec.Transitions, target)
}

// IsTerminal reports whether no further transition is possible.
func (m StateMachine) IsTerminal(status Status) bool {
	return slices.Contains(m.Terminal, status)
}

// Owner is the side responsible for moving a run out of status.
func (m StateMachine) Owner(status Status) string { return m.States[status].Owner }

func (m StateMachine) validate() error {
	for _, status := range AllStatuses {
		if _, ok := m.States[status]; !ok {
			return fmt.Errorf("run_states.yaml: state %q missing", status)
		}
	}
	if len(m.States) != len(AllStatuses) {
		return fmt.Errorf("run_states.yaml: %d states, Go knows %d", len(m.States), len(AllStatuses))
	}
	for _, status := range m.Terminal {
		if len(m.States[status].Transitions) > 0 {
			return fmt.Errorf("run_states.yaml: terminal state %q has transitions", status)
		}
	}
	if m.MaxAttempts < 1 {
		return fmt.Errorf("run_states.yaml: max_attempts must be >= 1")
	}
	return nil
}

// ErrorCode is one entry of contracts/error_codes.yaml.
type ErrorCode struct {
	Category    string `yaml:"category"`
	Expected    bool   `yaml:"expected"`
	Retryable   bool   `yaml:"retryable"`
	Description string `yaml:"description"`
}

// Contracts bundles both files.
type Contracts struct {
	States StateMachine
	Codes  map[string]ErrorCode
}

// Load reads and cross-checks both contract files.
func Load(dir string) (Contracts, error) {
	var out Contracts
	if err := readYAML(filepath.Join(dir, "run_states.yaml"), &out.States); err != nil {
		return out, err
	}
	if err := out.States.validate(); err != nil {
		return out, err
	}
	var codes struct {
		Codes map[string]ErrorCode `yaml:"codes"`
	}
	if err := readYAML(filepath.Join(dir, "error_codes.yaml"), &codes); err != nil {
		return out, err
	}
	out.Codes = codes.Codes
	for _, code := range CodesUsed {
		if _, ok := out.Codes[code]; !ok {
			return out, fmt.Errorf("error_codes.yaml: the dispatcher uses %s, which is not defined", code)
		}
	}
	return out, nil
}

func readYAML(path string, target any) error {
	raw, err := os.ReadFile(path)
	if err != nil {
		return fmt.Errorf("read %s: %w", filepath.Base(path), err)
	}
	if err := yaml.Unmarshal(raw, target); err != nil {
		return fmt.Errorf("parse %s: %w", filepath.Base(path), err)
	}
	return nil
}
